app_dart/lib/src/model/firestore/ci_staging.dart

```diff
@@ -108,49 +121,109 @@ class CiStaging extends Document {
+        transaction: transaction,
+      );
 
       final fields = doc.fields;
       if (fields == null) {
         throw '$logCrumb: missing fields for $transaction / $doc';
       }
 
       // Fields and remaining _must_ be present.
-      final remaining = int.tryParse(fields[kRemainingField]?.integerValue ?? '');
-      if (remaining == null) {
+      final docRemaining = int.tryParse(fields[kRemainingField]?.integerValue ?? '');
+      if (docRemaining == null) {
         throw '$logCrumb: missing field "$kRemainingField" for $transaction / ${doc.fields}';
       }
+      remaining = docRemaining;
+
+      final maybeFailed = int.tryParse(fields[kFailedField]?.integerValue ?? '');
+      if (maybeFailed == null) {
+        throw '$logCrumb: missing field "$kFailedField" for $transaction / ${doc.fields}';
+      }
+      failed = maybeFailed;
 
       // We will have check_runs scheduled after the engine was built successfully, so missing the checkRun field
       // is an OK response to have. All fields should have been written at creation time.
-      if (fields[checkRun] == null || fields[checkRun]!.stringValue == null) {
+      final recordedConclusion = fields[checkRun]?.stringValue;
+      if (recordedConclusion == null) {
         log.info('$logCrumb: $checkRun not present in doc for $transaction / $doc');
         await docRes.rollback(RollbackRequest(transaction: transaction), kDatabase);
-        return (valid: false, remaining: remaining);
+        return (valid: false, remaining: remaining, checkRunGuard: null, failed: failed);
       }
 
-      // Guard against going negative and log enough info so we can debug.
-      if (remaining == 0) {
-        throw '$logCrumb: field "$kRemainingField" is already zero for $transaction / ${doc.fields}';
-      }
-      newRemaining = remaining - 1;
-
-      // Choose what to update in the document; by default we want to record the new conclusion, but we only
-      // want to record `remaining - 1` if the original conclusion was `scheduled`.
-      if (kDefaultTaskStatus == fields[checkRun]!.stringValue) {
-        log.info('$logCrumb: setting remaining to $newRemaining and changing ${fields[checkRun]!.stringValue}');
-        fields[checkRun] = Value(stringValue: conclusion);
-        fields[kRemainingField] = Value(integerValue: '$newRemaining');
+      // GitHub sends us 3 "action" messages for check_runs: created, completed, or rerequested.
+      //   - We are responsible for the "created" messages.
+      //   - The user is responsible for "rerequested"
+      //   - LUCI is responsible for the completed.
+      // Completed messages are either success / failure.
+      // "remaining" should only go down if the previous state was scheduled - this is the first state
+      // that is written by the scheduler.
+      // "failed_count" can go up or down depending on:
+      //   previous == failure && current == success: +1
+      //   previous != failure && current == failure: -1
```

Comment:
If "previous" refers to "recordedConclusion" and "current" refers to "conclusion", then shouldn't the +1 and -1 be swapped in the comment above?

Reply:
yes - I screwed this up when I was renaming variables. Thanks for the catch!

Reply:
done